hbase-server/src/test/java/org/apache/hadoop/hbase/master/assignment/TestExceptionInAssignRegion.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.hadoop.hbase.util.JVMClusterUtil;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category({ MasterTests.class, MediumTests.class })
+public class TestExceptionInAssignRegion {
+
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestExceptionInAssignRegion.class);
+
+  private static final HBaseTestingUtility UTIL = new HBaseTestingUtility();
+
+  private static final TableName TABLE_NAME = TableName.valueOf("test");
+
+  private static final CountDownLatch countDownLatch = new CountDownLatch(2);
+
+  private static final byte[] CF = Bytes.toBytes("cf");
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    UTIL.getConfiguration().setStrings(CoprocessorHost.REGION_COPROCESSOR_CONF_KEY,
+      ThrowInOpenCP.class.getName());
+    UTIL.startMiniCluster(3);
+    UTIL.getAdmin().balancerSwitch(false, true);
+    UTIL.createTable(TABLE_NAME, CF);
+    UTIL.waitTableAvailable(TABLE_NAME);
+  }
+
+  @AfterClass
+  public static void tearDown() throws Exception {
+    UTIL.shutdownMiniCluster();
+  }
+
+  @Test
+  public void testExceptionInAssignRegion() {
+    ProcedureExecutor procedureExecutor =
+      UTIL.getMiniHBaseCluster().getMaster().getMasterProcedureExecutor();
+
+    JVMClusterUtil.RegionServerThread rsThread = null;
+    for (JVMClusterUtil.RegionServerThread t : UTIL.getMiniHBaseCluster()
+      .getRegionServerThreads()) {
+      if (!t.getRegionServer().getRegions(TABLE_NAME).isEmpty()) {
+        rsThread = t;
+        break;
+      }
+    }
+    // find the rs and hri of the table
+    HRegionServer rs = rsThread.getRegionServer();
+    RegionInfo hri = rs.getRegions(TABLE_NAME).get(0).getRegionInfo();
+    TransitRegionStateProcedure assignRegionProcedure = TransitRegionStateProcedure.move(
+      UTIL.getMiniHBaseCluster().getMaster().getMasterProcedureExecutor().getEnvironment(),
+      hri, null);
+    RegionStateNode regionNode = UTIL.getMiniHBaseCluster().getMaster().getAssignmentManager()
+      .getRegionStates().getOrCreateRegionStateNode(hri);
+    regionNode.setProcedure(assignRegionProcedure);
+    countDownLatch.countDown();
+    long prodId = procedureExecutor.submitProcedure(assignRegionProcedure);
+    ProcedureTestingUtility.waitProcedure(procedureExecutor, prodId);
+
+    Assert.assertEquals("Should be two RS since other is aborted", 2,
+      UTIL.getMiniHBaseCluster().getLiveRegionServerThreads().size());
+    Assert.assertNull("RIT Map doesn't have correct value",
+      getRegionServer(0).getRegionsInTransitionInRS().get(hri.getEncodedNameAsBytes()));
+    Assert.assertNull("RIT Map doesn't have correct value",
+      getRegionServer(1).getRegionsInTransitionInRS().get(hri.getEncodedNameAsBytes()));
```

Comment:
not sure how do you make sure that region transition won't finish until you run this test here?

Reply:
any thought on this?

Reply:
@xcangCRM  we wait for the procedure to complete.